Data/Sequence/Internal.hs

```diff
@@ -4604,7 +4607,35 @@ sortBy cmp (Seq xs) =
     maybe
         (Seq EmptyT)
         (execState (replicateA (size xs) (popMinS cmp)))
-        (toPQS cmp (Seq xs))
+        (toPQS cmp (\s (Elem x) -> PQS s x Nl) 0 xs)
+
+-- | \( O(n \log n) \). 'sortOn' sorts the specified 'Seq' by comparing
+-- the results of a key function applied to each element. @'sortOn' f@ is
+-- equivalent to @'sortBy' ('compare' ``Data.Function.on`` f)@, but has the
+-- performance advantage of only evaluating @f@ once for each element in the
+-- input list. This is called the decorate-sort-undecorate paradigm, or
+-- Schwartzian transform.
+--
+-- An example of using 'sortOn' might be to sort a 'Seq' of strings
+-- according to their length:
+--
+-- > sortOn length (fromList ["alligator", "monkey", "zebra"]) == fromList ["zebra", "monkey", "alligator"]
+--
+-- If, instead, 'sortBy' had been used, 'length' would be evaluated on
+-- every comparison, giving \( O(n \log n) \) evaluations, rather than
+-- \( O(n) \).
+--
+-- If @f@ is very cheap (for example a record selector, or 'fst'),
+-- @'sortBy' ('compare' ``Data.Function.on`` f)@ will be faster than
+-- @'sortOn' f@.
+sortOn :: Ord b => (a -> b) -> Seq a -> Seq a
+sortOn f (Seq xs) =
+    maybe
+       (Seq EmptyT)
+       (execState (replicateA (size xs) (fmap (\(_ :*: x) -> x) (popMinS compareFst))))
+       (toPQS compareFst (\s (Elem x) -> PQS s (f x :*: x) Nl) 0 xs)
+  where
```

Comment:
I think we probably want to use yet another type instead of reusing `PQS`, to reduce the allocation and indirection of the strict pair constructor. Sorry; I know it's a bit annoying.

Reply:
No problem, for the normal PQueue as well?

Reply:
Actually, if that means there's going to be 4 different queues, it might begin to make sense to add functions like:

```haskell
foldDigit :: (b -> b -> b) -> (a -> b) -> Digit a -> b
foldDigit (<+>) = fm where
    fm f (One a) = f a
    fm f (Two a b) = f a <+> f b
    fm f (Three a b c) = f a <+> f b <+> f c
    fm f (Four a b c d) = f a <+> f b <+> f c <+> f d
{-# INLINE foldDigit #-}
```

and

```haskell
{-# SPECIALIZE foldWithIndexDigit :: (b -> b -> b) -> (Int -> Elem y -> b) -> Int -> Digit (Elem y) -> b #-}
{-# SPECIALIZE foldWithIndexDigit :: (b -> b -> b) -> (Int -> Node y -> b) -> Int -> Digit (Node y) -> b #-}
foldWithIndexDigit :: Sized a => (b -> b -> b) -> (Int -> a -> b) -> Int -> Digit a -> b
foldWithIndexDigit _ f !s (One a) = f s a
foldWithIndexDigit (<+>) f s (Two a b) = f s a <+> f sPsa b
  where
    !sPsa = s + size a
foldWithIndexDigit (<+>) f s (Three a b c) = f s a <+> f sPsa b <+> f sPsab c
  where
    !sPsa = s + size a
    !sPsab = sPsa + size b
foldWithIndexDigit (<+>) f s (Four a b c d) =
    f s a <+> f sPsa b <+> f sPsab c <+> f sPsabc d
  where
    !sPsa = s + size a
    !sPsab = sPsa + size b
    !sPsabc = sPsab + size c
```

And use them in the `toPQ` functions, as well as `foldMap` and `foldMapWithIndex`. What do you think?